src/utils/changelog.ts

```diff
@@ -642,11 +664,26 @@ function formatChangelogEntry(entry: ChangelogEntry): string {
   return text;
 }
 
+/**
+ * Result of changelog generation, includes both the formatted changelog
+ * and the determined version bump type based on commit categories.
+ */
+export interface ChangelogResult {
+  /** The formatted changelog string */
+  changelog: string;
+  /** The highest version bump type found, or null if no commits matched categories with semver */
+  bumpType: BumpType | null;
+  /** Number of commits analyzed */
+  totalCommits: number;
+  /** Number of commits that matched a category with a semver field */
+  matchedCommitsWithSemver: number;
+}
+
 export async function generateChangesetFromGit(
   git: SimpleGit,
   rev: string,
   maxLeftovers: number = MAX_LEFTOVERS
-): Promise<string> {
+): Promise<ChangelogResult> {
```

Comment:
Instead of passing this function's result around for caching, we should memoize this function. This should be easy as none of the arguments (`git`, `rev`, `maxLeftovers`) can be changed throughout a single run (they are the inputs and we don't do multiple passes on different repos etc). Still use proper memoization with taking argument values into account.

Since we are returning a promise, we should memoize the promise so any concurrent calls while the promise is still pending get coalesced.